lib/datasets.go

```diff
@@ -20,33 +20,66 @@ import (
 	"github.com/qri-io/qri/rev"
 )
 
-// DatasetRequests encapsulates business logic for working with Datasets on Qri
-type DatasetRequests struct {
-	cli  *rpc.Client
-	node *p2p.QriNode
+// DatasetMethods encapsulates all functions for working with datasets, composing
+// dataset read, write, and plumbing operations
+type DatasetMethods interface {
+	Methods
+	DatasetReadMethods
+	DatasetWriteMethods
+	DAGMethods
+}
+
+// DatasetReadMethods defines all read functions that operate on a dataset
+type DatasetReadMethods interface {
+	List(p *ListParams, res *[]repo.DatasetRef) error
+	Get(p *GetParams, res *GetResult) error
+}
+
+// PublishMethods
+type PublishMethods interface {
+	SetPublishStatus(p *SetPublishStatusParams, publishedRef *repo.DatasetRef) error
+}
+
+// DatasetWriteMethods defines all read functions that operate on a dataset
+type DatasetWriteMethods interface {
+	Save(p *SaveParams, res *repo.DatasetRef) error
+	Rename(p *RenameParams, res *repo.DatasetRef) error
+	Remove(p *RemoveParams, res *RemoveResponse) error
+	Add(ref, res *repo.DatasetRef) error
 }
 
-// CoreRequestsName implements the Requets interface
-func (DatasetRequests) CoreRequestsName() string { return "datasets" }
+// DAGMethods implements
+// TODO (b5): this whole interface really doesn't belong on to dataset stuff
+// it's generic to DAGs
+type DAGMethods interface {
+	Manifest(refstr *string, m *dag.Manifest) (err error)
+	ManifestMissing(a, b *dag.Manifest) (err error)
+	DAGInfo(s *DAGInfoParams, i *dag.Info) (err error)
+}
 
-// NewDatasetRequests creates a DatasetRequests pointer from either a repo
+// NewDatasetMethods creates a DatasetMethods pointer from either a repo
 // or an rpc.Client
-func NewDatasetRequests(node *p2p.QriNode, cli *rpc.Client) *DatasetRequests {
-	if node != nil && cli != nil {
-		panic(fmt.Errorf("both repo and client supplied to NewDatasetRequests"))
+func NewDatasetMethods(inst Instance) DatasetMethods {
+	return datasetMethods{
+		node: inst.Node(),
+		cli:  inst.RPC(),
 	}
+}
 
-	return &DatasetRequests{
-		node: node,
-		cli:  cli,
-	}
+// datasetMethods encapsulates business logic for working with Datasets on Qri
+type datasetMethods struct {
+	cli  *rpc.Client
```

Comment:
Nearly every *Methods struct is going to be wrapping the same state: an rpc.Client, a p2p.Node, and a Config. I think it's worth wrapping them together into a single object that we can pass around a pointer to. Call it MethodState or OperationState or something. Then, these *Methods structs are thin encapsulations over this state, along with a bit of info for method dispatch.

Not a huge fan of Capital Interface with lower-case implementation. Makes it hard to speak aloud about. Maybe DatasetMethods for interface, and datasetOperations / datasetOps for internal struct?

Reply:
Deep agree here. So far I've been calling that encapsulated state an `Instance`. This PR adjusts method constructor func signatures to accept only an Instance. eg: 
```go
func NewDatasetMethods(inst Instance) DatasetMethods
```

So far I'm liking the way the phrasing shakes out: "derive the methods you need from an instance".

As for the exported-interface unexported-implementation, I will admit to being a huge fan. I've seen it in other codebases (examples escape me at the moment, I think it might have been a stats package...). The way I verbalize it is there's the _Interface_, and a _default implementation_. IMHO the reader can clearly see the intent of the private struct: implement the interface.